04-Protocol-Security.md

```diff
@@ -36,32 +36,40 @@ These primitives are chosen so that Noise Encryption layer for Stratum V2 can be
 
 #### 4.3.1.1 EC point encoding remarks
 
-Secp256k1 curve points, which includes Public Keys and ECDH results, are points with of X- and Y-coordinate, 32-bytes each. There are several possibilities how to serialize them:
+Secp256k1 curve points, i.e. Public Keys, are points with of X- and Y-coordinate.
+We serialize them in three different ways, only using the x-coordinate.
 
-1. 64-byte full X- and Y-coordinate serialization for public keys (and ECDH results) and 96 bytes for signatures.
-2. 33-byte X-coordinate with 1 parity bit serialization for public keys and similarly 65-byte for signatures.
-3. 32-byte X-coordinate only with implicit Y-coordinate for public keys and 64-byte for signatures.
+1. When signing or verifying a certificate, we use the 32 byte x-only
+encoding as defined in BIP 340.<sup>[3](#reference-3)</sup>.
 
-We choose the 32-byte serialization for public key and 64-byte for signatures with implicit Y-coordinate.
+2. When sharing keys during the handshake, whether in plain text or encrypted,
+we use the 64 byte ElligatorSwift x-only encoding as defined in BIP324<sup>[7](#reference-7)</sup> under "ElligatorSwift encoding of curve X coordinates". This encoding uses 64-bytes instead of 32-bytes in order to produce a
+pseudo-random bytesteam. This is useful because the protocol handshake starts with
+each side sending ther public key in plain text. Additionally the use of X-only
+ElligatorSwift ECDH removes the need to grind or negate private keys.
 
-The parity of Y-coordinate is always assumed to be even.
+3. The Authority public key is base58-check encoded as described in 4.7.
+
+Digital signatures are serialized in 64-bytes like in BIP340<sup>[3](#reference-3)</sup>.
 
 Key generation algorithm:
 
 1. generate random 32-byte secret key `sk`
 2. let `d' = int(sk)`
 3. fail if `d = 0` or `d' > n` where `n` is group order of secp256k1 curve
 4. compute `P` as `d'⋅G`
-5. drop the Y coordinate and output keypair `(sk, bytes(P.x))`
-
-Such system has the following properties:
+5. drop the Y coordinate and compute (u, t) = XElligatorSwift(P.x)
```

Comment:
I'm slightly confused. Are there two types of ElligatorSwift? Normal and X(only)Elligator?

the rust library secp256k1 offers only API to construct elligator from full public key.

Reply:
The function name `XElligatorSwift` comes from BIP324. I use `XElligatorSwift(P.x)` instead of `XElligatorSwift(x)` to stay closer to the original notation in the sv2 spec.

`secp256k1`, the rust binding, as well as the Bitcoin Core bindings take care of many steps described in the BIP, so that's not really a contradiction. E.g. in Bitcoin Core you first have to convert a regular `CPubKey` to an `XOnlyPubKey` and only then to an `EllSwiftPubKey`.

Reply:
Fair enough.

I run some test just to make sure:
```rust
        use secp256k1::ellswift::*;
        let key_pair_a = SimpleKeyGen.new_keypair();
        let xpk_a = key_pair_a.pk;

        let key_pair_b = SimpleKeyGen.new_keypair();
        let xpk_b = key_pair_b.pk;


        let ellswift_a_even = ElligatorSwift::from_pubkey(PublicKey::from_x_only_public_key(xpk_a, secp256k1::Parity::Even));
        let ellswift_a_odd = ElligatorSwift::from_pubkey(PublicKey::from_x_only_public_key(xpk_a, secp256k1::Parity::Even));

        let ellswift_b_even =ElligatorSwift::from_pubkey(PublicKey::from_x_only_public_key(xpk_b, secp256k1::Parity::Even));
        let ellswift_b_odd = ElligatorSwift::from_pubkey(PublicKey::from_x_only_public_key(xpk_b, secp256k1::Parity::Even));


        let x1 = *ElligatorSwift::shared_secret(ellswift_a_even, ellswift_b_even, key_pair_a.sk, ElligatorSwiftParty::A, None).as_secret_bytes();
        let x2 = *ElligatorSwift::shared_secret(ellswift_a_odd, ellswift_b_even, key_pair_a.sk, ElligatorSwiftParty::A, None).as_secret_bytes();
        let x3 = *ElligatorSwift::shared_secret(ellswift_a_even, ellswift_b_odd, key_pair_a.sk, ElligatorSwiftParty::A, None).as_secret_bytes();
        let x4 = *ElligatorSwift::shared_secret(ellswift_a_odd, ellswift_b_odd, key_pair_a.sk, ElligatorSwiftParty::A, None).as_secret_bytes();


        let x5 = *ElligatorSwift::shared_secret(ellswift_a_even, ellswift_b_even, key_pair_b.sk, ElligatorSwiftParty::B, None).as_secret_bytes();
        let x6 = *ElligatorSwift::shared_secret(ellswift_a_odd, ellswift_b_even, key_pair_b.sk, ElligatorSwiftParty::B, None).as_secret_bytes();
        let x7 = *ElligatorSwift::shared_secret(ellswift_a_even, ellswift_b_odd, key_pair_b.sk, ElligatorSwiftParty::B, None).as_secret_bytes();
        let x8 = *ElligatorSwift::shared_secret(ellswift_a_odd, ellswift_b_odd, key_pair_b.sk, ElligatorSwiftParty::B, None).as_secret_bytes();

        assert_eq!(x1, x2);
        assert_eq!(x1, x3);
        assert_eq!(x1, x4);
        assert_eq!(x1, x5);
        assert_eq!(x1, x6);
        assert_eq!(x1, x7);
        assert_eq!(x1, x8);
```

It passed.

Reply:
@Sjors Apology, I made a mistake. I forgot to change the parity.

This test actually **doesn't pass**

```rust
use secp256k1::ellswift::*;
let key_pair_a = SimpleKeyGen.new_keypair();
let xpk_a = key_pair_a.pk;

let key_pair_b = SimpleKeyGen.new_keypair();
let xpk_b = key_pair_b.pk;

let ellswift_a_even = ElligatorSwift::from_pubkey(PublicKey::from_x_only_public_key(xpk_a, secp256k1::Parity::Even));
let ellswift_a_odd = ElligatorSwift::from_pubkey(PublicKey::from_x_only_public_key(xpk_a, secp256k1::Parity::Odd));

let ellswift_b_even =ElligatorSwift::from_pubkey(PublicKey::from_x_only_public_key(xpk_b, secp256k1::Parity::Even));


let x1 = *ElligatorSwift::shared_secret(ellswift_a_even, ellswift_b_even, key_pair_a.sk, ElligatorSwiftParty::A, None).as_secret_bytes();
let x2 = *ElligatorSwift::shared_secret(ellswift_a_odd, ellswift_b_even, key_pair_a.sk, ElligatorSwiftParty::A, None).as_secret_bytes();

assert_eq!(x1, x2);
```

Reply:
From my experiments it appears as Elligator, as provided by the rust secp256k1 library, does depend on the key parity.
However the raw elligator ECDH output does not.

The following test indeed passes.

```rust
use secp256k1::ellswift::*;
use secp256k1::*;

let key_pair_a = SimpleKeyGen.new_keypair();
let xpk_a = key_pair_a.pk;

let key_pair_b = SimpleKeyGen.new_keypair();
let xpk_b = key_pair_b.pk;

let ctx = Secp256k1::new();

let elligator_a_even = ElligatorSwift::from_pubkey(PublicKey::from_x_only_public_key(xpk_a, Parity::Even));
let elligator_a_odd = ElligatorSwift::from_pubkey(PublicKey::from_x_only_public_key(xpk_a, Parity::Odd));

let elligator_b_even = ElligatorSwift::from_pubkey(PublicKey::from_x_only_public_key(xpk_b, Parity::Even));

let hasher = |dh, a, b| {
    ElligatorSwiftSharedSecret::from_secret_bytes(dh)
};

let x1 = ElligatorSwift::shared_secret_with_hasher(elligator_a_even, elligator_b_even, key_pair_a.sk, ElligatorSwiftParty::A, hasher).to_secret_bytes();
let x2 = ElligatorSwift::shared_secret_with_hasher(elligator_a_odd, elligator_b_even, key_pair_a.sk, ElligatorSwiftParty::A, hasher).to_secret_bytes();

assert_eq!(x1, x2);
```

Reply:
I think what's supposed to work is when you negate `key_pair_a` / `key_pair_b`.

Reply:
> I think what's supposed to work is when you negate `key_pair_a` / `key_pair_b`. In your current test `key_pair_a.sk` is no longer the private key that corresponds with `ellswift_a_even` / `ellswift_a_odd`.

Negating the key doesn't help either. That's probably due to the hashing function covering the result of the "odd elligator".

Anyways, in my current understanding, Elligator technology itself isn't limited only to the X-coordinates. The Elligator *might* be used in a way that is consistent with x-only EC arithmetic, but it should be explicitly described how to do that. In our case: **always assume even parity**.
That BTW means, **never construct elligator from secret key directly unless you know the corresponding pubkey is even**

Reply:
> Using the full v2_ecdh would actually introduce the dependency on the key parity.

I don't believe this is true. BIP324 uses `v2_ecdh` and does not negate keys. I think the snippet above is not realistic; it's modifying the public key in a way that it no longer matches the private key. The EllSwift ECDH function takes both the private key and public key as argument, and assumes they weren't modified.

Reply:
Also note that `ellswift_xdh_hash_function_bip324` in libsecp _only_ covers the x-coordinate:

```c
static int ellswift_xdh_hash_function_bip324(unsigned char* output, const unsigned char *x32, const unsigned char *ell_a64, const unsigned char *ell_b64, void *data) {
    secp256k1_sha256 sha;

    (void)data;

    secp256k1_ellswift_sha256_init_bip324(&sha);
    secp256k1_sha256_write(&sha, ell_a64, 64);
    secp256k1_sha256_write(&sha, ell_b64, 64);
    secp256k1_sha256_write(&sha, x32, 32);
    secp256k1_sha256_finalize(&sha, output);

    return 1;
}
```

`ell_a64` and `ell_b64` are the 64 byte ellswift encoded public keys. They are produced by `ellswift_create` which only uses the x-coordinate (via XElligatorSwift).

What happens when you run the above test with `ellswift_ecdh_xonly`?

Reply:
> > Using the full v2_ecdh would actually introduce the dependency on the key parity.
> 
> I don't believe this is true. BIP324 uses `v2_ecdh` and does not negate keys. I think the snippet above is not realistic; it's modifying the public key in a way that it no longer matches the private key. The EllSwift ECDH function takes both the private key and public key as argument, and assumes they weren't modified.

You are right. Sorry, I was quite confused when testing and learning all the properties.

In the end I arrived to the conclusion, that elligator indeed contains pubkey parity information, but it is irrelevant to the standard use case scenario. The aim is to exchange elligators and perform ECDH using one of the party's private key.
The private key sign doesn't have an impact on the ECDH result - either on `v2_ecdh` or on `ellswift_ecdh_xonly` and that's important. That's why your code works as you mentioned above.

The algorithm is not required to be invariant under the elligator transformation (e. g. extracting public key and transforming it back to an elligator with a different key-parity).